text/0000-js-compatible-resolution.md

```diff
@@ -0,0 +1,344 @@
+
+## Auto-insertion of imports
+
+You don't need to import built-in constructs that ship with Ember (like `if`) or Ember's built-in helpers (like `action`).
+
+To maintain framework extensibility and prevent these built-ins from being so "magical" and privileged, we offer a clear path for apps to register similar global names:
+
+```js
+//config/template-globals.js
+export default {
+  t: "import { t } from 'ember-intl'"
+};
+```
+
+This says "whenever the template compiler encounters the undefined name `t`, it should automatically insert `"import { t } from 'ember-intl'"` at the start of the template. This only applies to otherwise _undefined_ names: if a template creates its own binding for `t`, that takes precedence.
+
+Addons are deliberately _not_ allowed to register themselves into the template-globals. They can offer a generator that helps app-authors edit `template-globals.js`, but control over globals remains _explicit in the app_, so that there's only one place where you need to look to find where a global name is coming from.
+
+## Rationalizing Components as Modules
+
+A key piece of this design is explaining how to interpret Ember components as Javascript modules. There are some important requirements:
+
+ - When a component has a Javascript module, and you import the component from Javascript, you should get the values back you'd expect for a normal Javascript module. We don't want to rewrite things in an opaque way.
+ - Consumers of a component should not need to know whether it is implemented as only a template, a template plus Javascript, or only Javascript.
+ - A component that is reexported should continue to work unchanged (in contrast with today, in which component authors needs to take special care to make their templates come along).
+ - We want to follow the _existing_ Module Resolver rules and not introduce new ones. Especially new ones that would diverge further from typical node_modules resolution.
+
+Given these requirements, the proposed design is:
+
+1. Imports in templates are resolved the same way they are resolved today in Javascript. Conceptually, we add a custom handler for `.hbs` files that allows them to be resolved the same as `.js` files. `.js` files take precedence over `.hbs` files, such that if both exist for a given ModuleSpecifier, the JS wins.
+
+2. Component Javascript should be authored using default exports (as they already are):
+
+    ```js
+    export default class extends Component {
+      yourStuffHere(){}
+    };
+    ```
+
+    There is no magic that alters the exported value. When it's imported (whether in Javacript or HBS), the value you get is the actual class.
+
+3. When both `.js` and `.hbs` exist for a component, the build system unobtrusively associates the template with the default value exported by the Javascript. The template must be co-located with the Javascript, having the same filename except for changing the extension to `.hbs`.
```

Comment:
I totally dig this in principle, but there are specific edge cases that we need to make sure are handled properly when we actually implement. For example, `Ember.Component`'s today can have a different `.layout` property _per instance_. I definitely do not think this is a blocker at all, but likely means that we need to add a `getLayout` / `getTemplate` capability to the component manager interface. In `Ember.Component` this method would have to look at `component.layout` first then fall back to this loose/unobtrusive association.

Reply:
I'd like to see the detailed design specifically call out the `getLayout` capability I mentioned above.

Reply:
> The template must be co-located with the Javascript, having the same filename except for changing the extension to `.hbs`.

This would expect something like:

```
src/ui/components/
  async-button.js
  async-button.hbs
  title-bar.js
  title-bar.hbs
```

Do you not mean to support this?

```
src/ui/components/
  async-button/
    component.js
    template.hbs
  title-bar/
    component.js
    template.hbs
```

Reply:
https://github.com/ef4/rfcs/blob/2d4e4a46dd9bd9903aa5d59f230b765c4dd163d6/text/0000-js-compatible-resolution.md#L191-L194

Taking the above into account, it looks like it. This has implications for the MU resolver regarding the default types for collections, right?

Reply:
https://github.com/ef4/rfcs/blob/2d4e4a46dd9bd9903aa5d59f230b765c4dd163d6/text/0000-js-compatible-resolution.md#L271-L277

aha!